code/core_io_tree.c

```diff
@@ -11,42 +11,37 @@
 #include "core_allvars.h"
 #include "core_proto.h"
 
+#include "io/tree_binary.h"
+#ifdef HDF5
+#include "io/tree_hdf5.h"
+#endif
 
-// keep a static file handle to remove the need to do constant seeking
-FILE* load_fd = NULL;
-
-
-void load_tree_table(int filenr)
+void load_tree_table(int filenr, enum Valid_TreeTypes my_TreeType)
 {
-  int i, n, totNHalos;
-  char buf[1000];
+  int i, n;
   FILE *fd;
+  char buf[MAX_STRING_LEN];
 
-	// open the file each time this function is called
-  sprintf(buf, "%s/%s.%d", SimulationDir, TreeName, filenr);
-  if(!(load_fd = fopen(buf, "r")))
+  switch (my_TreeType)
   {
-    printf("can't open file `%s'\n", buf);
-    ABORT(0);
+#ifdef HDF5
+    case genesis_lhalo_hdf5:
+      load_tree_table_hdf5(filenr);
+      break;
+#endif
+      
+    case lhalo_binary:
+      load_tree_table_binary(filenr);
+      break;
+
+    default :
+      fprintf(stderr, "The specified TreeType was my_TreeType\nThis is not a valid option.");
+      ABORT(0);
   }
 
-  myfread(&Ntrees, 1, sizeof(int), load_fd);
-  myfread(&totNHalos, 1, sizeof(int), load_fd);
-
-  TreeNHalos = mymalloc(sizeof(int) * Ntrees);
-  TreeFirstHalo = mymalloc(sizeof(int) * Ntrees);
-
-  for(n = 0; n < NOUT; n++)
-    TreeNgals[n] = mymalloc(sizeof(int) * Ntrees);
-  myfread(TreeNHalos, Ntrees, sizeof(int), load_fd);
-
-  if(Ntrees)
-    TreeFirstHalo[0] = 0;
-  for(i = 1; i < Ntrees; i++)
-    TreeFirstHalo[i] = TreeFirstHalo[i - 1] + TreeNHalos[i - 1];
-
   for(n = 0; n < NOUT; n++)
   {
+    TreeNgals[n] = mymalloc(sizeof(int) * Ntrees);
```

Comment:
For the future, we should replace `mymalloc` to be something like `calloc`. Removes any potential overflows that can occur. For now, as long as we are compiling on 64bit platforms, `size_t` should be `uint64_t` and the overflow should not occur in the multiplication (since `Ntrees` gets promoted to `uint64_t` and the multiplication occurs between two `uint64_t`)